arwen/contexts/async_local.go

```diff
@@ -0,0 +1,454 @@
+// built-in function. This is required for all cross-shard calls to built-in
+// functions, because they will handle cross-shard calls themselves, by
+// generating entries in vmOutput.OutputAccounts, and they need to be executed
+// synchronously to do that. As a consequence, it is not necessary to call
+// sendAsyncCallCrossShard(). The vmOutput produced by the built-in function,
+// containing the cross-shard call, has ALREADY been merged into the main
+// output by the inner call to host.ExecuteOnDestContext(). Moreover, the
+// status of the AsyncCall is not updated here - it will be updated by
+// PostprocessCrossShardCallback(), when the cross-shard call returns.
+func (context *asyncContext) executeSyncHalfOfBuiltinFunction(asyncCall *arwen.AsyncCall) error {
+	destinationCallInput, err := context.createContractCallInput(asyncCall)
+	if err != nil {
+		return err
+	}
+
+	// Briefly restore the AsyncCall GasLimit, after it was consumed in its
+	// entirety by addAsyncCall(); this is required, because ExecuteOnDestContext()
+	// must also consume the GasLimit in its entirety, before starting execution,
+	// but will restore any GasRemaining to the current instance.
+	metering := context.host.Metering()
+	metering.RestoreGas(asyncCall.GetGasLimit())
+
+	vmOutput, err := context.host.ExecuteOnDestContext(destinationCallInput)
+	if err != nil {
+		return err
+	}
+
+	// If the in-shard half of the built-in function call has failed, go no
+	// further and execute the error callback of this AsyncCall.
+	if vmOutput.ReturnCode != vmcommon.Ok {
+		asyncCall.Reject()
+		callbackVMOutput, callbackErr := context.executeSyncCallback(asyncCall, vmOutput, err)
+		context.finishAsyncLocalExecution(callbackVMOutput, callbackErr)
+	}
+
+	// The gas that remains after executing the in-shard half of the built-in
+	// function is provided to the cross-shard half.
+	asyncCall.GasLimit = vmOutput.GasRemaining
+
+	return nil
+}
+
+// executeSyncContextCallback will execute the callback of the original caller
+// synchronously, already assuming the original caller is in the same shard
+func (context *asyncContext) executeSyncContextCallback() {
+	callbackCallInput := context.createContextCallbackInput()
+	callbackVMOutput, callBackErr := context.host.ExecuteOnDestContext(callbackCallInput)
+	context.finishAsyncLocalExecution(callbackVMOutput, callBackErr)
+}
+
+// TODO return values are never used by code that calls finishAsyncLocalExecution
+func (context *asyncContext) finishAsyncLocalExecution(vmOutput *vmcommon.VMOutput, err error) {
+	if err == nil {
+		return
+	}
+
+	runtime := context.host.Runtime()
+	output := context.host.Output()
+
+	runtime.GetVMInput().GasProvided = 0
+
+	if vmOutput == nil {
+		vmOutput = output.CreateVMOutputInCaseOfError(err)
+	}
+
+	// TODO Discuss consistency between in-shard and cross-shard results
+	// TODO of the callback, and how they're accessible to the caller / user.
```

Comment:
the returnCode cross shard is set. so we need to set inshard as well. We have to make the callBack aware that one call was wrong.

Reply:
Deleted `TODO` and uncommented the line after it.